Drivers: hv: mshv_vtl: Speed up PVALIDATE and RPMADJUST to meet perf …

romank-msft opened the pull request:
…targets

Using 4K pages for pvalidate breaks large VMs in the cloud as that
is very slow. Using 2M pages for mapping arbitrary PFNs (i.e. not
having a tracking struct page*) isn't obvious in the kernel.

Employ the VA gap documented as meant for the hypervisor use for creating local maps.
These are local to the processor and are used to briefly map the PFN requested to
be pvalidate-d or rmpadjust-ed.

The implementation allows to parallelize the load between the CPUs if the user mode
code wishes to do so. The 5-level page tables are supported as well as 4-level ones.

The 64GiB+64VP VMs accept the memory in under a second on average with this change,
and without that that's ~57sec on average.

romank-msft commented:
1. Break out the if statements.
2. Change the init routine so there would be no need in updating user space

benhillis commented:
Could you please rebase this change? Sorry the dev branch got into a bad state because of an incorrect merge type the last time I took latest stable patches. Thanks!

romank-msft commented:
> Could you please rebase this change? Sorry the dev branch got into a bad state because of an incorrect merge type the last time I took latest stable patches. Thanks!

Thanks for the heads up, will do!

romank-msft commented:
> Could you please rebase this change? Sorry the dev branch got into a bad state because of an incorrect merge type the last time I took latest stable patches. Thanks!

done

allenpais commented:
@romank-msft  Could you please check the change again. there are conflicts.

romank-msft commented:
> @romank-msft Could you please check the change again. there are conflicts.

Will do, thanks for the heads up!

romank-msft commented:
> @romank-msft Could you please check the change again. there are conflicts.

done

allenpais commented:
> > @romank-msft Could you please check the change again. there are conflicts.
> 
> done

Thank you!. Could I get someone from your team to approve too before I go ahead with the merge.

romank-msft commented:
> > > @romank-msft Could you please check the change again. there are conflicts.
> > 
> > 
> > done
> 
> Thank you!. Could I get someone from your team to approve too before I go ahead with the merge.

Not ready yet for merge, got to tweak two tiny things mentioned after the commit description. Might get to that today after sending out v7 or the arm64 patches and conf VMBus (sending or preparing)

allenpais commented:
> > > > @romank-msft Could you please check the change again. there are conflicts.
> > > 
> > > 
> > > done
> > 
> > 
> > Thank you!. Could I get someone from your team to approve too before I go ahead with the merge.
> 
> Not ready yet for merge, got to tweak two tiny things mentioned after the commit description. Might get to that today after sending out v7 or the arm64 patches and conf VMBus (sending or preparing)

Sure, we'll work on it. Thanks.

namancse commented:
Hi Roman,
Since this change touches memory zones, watchdog petting logic, it would be good to have this thoroughly tested before merging on different VM SKUs. If you've already covered it, we are good.

namancse commented on drivers/hv/mshv_vtl_local_maps.c:
```diff
@@ -0,0 +1,528 @@
+ *
+ * Let `start` = `ffff800000000000` (PML4 index 256).
+ *
+ * 2MiB maps
```
What is the use case for having both regular and large page maps for each processor?

romank-msft replied:
The instructions thar the guest uses for (un)accepting pages offered by the hypervisor (`pvalidate`) and adjusting permissions on them (`rmpadjust`) might operate on both large and small pages. The page subject to (un)/accepting or adjusting permissions in the RMP (reverse map table) has to be mapped in the guest as the hardware requires. It is far more advantageous to use large pages but the memory map supplied by the host doesn't always allow that, usually due to fragmentation of the SPA (System Physical Address) space which the hypervisor uses to construct the GPA (Guest Physical Address) space.

Some pointers from the AMD PPR (Processor Programming Reference) Vol 2 "System Programming" available from the AMD website:
* `15.36.11 Large Page Management`
* `15.36.3 Reverse Map Table`

and from Vol3 loc. cit.:
* `4 System Instruction Reference`

namancse commented on drivers/hv/mshv_vtl_local_maps.c:
```diff
@@ -0,0 +1,528 @@
+ *
+ * Safety
+ * ======
+ * This kibd of code might be very dangerous to run, and there are many checks
```
When you were validating this, was there any way for you to test and confirm that this memory region (the one reserved for hypervisor) is actually not being used by anyone?

romank-msft replied:
The documentation claims it has to be free. The code checks that the page table entries are marked as non-present (the LSB of the PTE is 0). Of course that might be a special PTE marked for the backing with the phys page upon a page fault, and for that reason the code checks for the entire PTE to be `0`

namancse commented on drivers/hv/mshv_vtl_local_maps.c:
```diff
@@ -0,0 +1,528 @@
+ * and similar (where mapping to some/any VA is needed for the operation), should open
+ * and keep opened `/dev/mshv_local_maps` while these operations are in progress.
+ * Opening is idempotent hence opening several times doesn't change anything for the
+ * process. The processing happens per-CPU with preemption disabled (the watchdogs
```
What is the reason for disabling preemption?

romank-msft replied:
The instructions that's built for (`pvalidate` and `rmpadjust`) themselves don't need preemption disabled as they can run on any VP. The VA mappings are specific to the VPs to avoid remote TLB flushes and be able to run these instructions in parallel. As the VA mappings are specific to the VPs, the code has to stay pinned to the VP thus the preemption is disabled.

namancse commented on drivers/hv/mshv_vtl_local_maps.c:
```diff
@@ -0,0 +1,528 @@
+ * =====
+ * The process that wishes to pass PFNs to the kernel for processing with `pvalidate`
+ * and similar (where mapping to some/any VA is needed for the operation), should open
+ * and keep opened `/dev/mshv_local_maps` while these operations are in progress.
```
What happens if '/dev/mshv_local_maps' is not kept opened?
IMO mshv_vtl_local_maps_release gets called. Will we see a bootup failure in this case for openHCL?

romank-msft replied:
The failure might happen anytime during the VM life when the pages are requested to be (un)accepted or privilege adjusted.

namancse commented on drivers/hv/mshv_vtl_local_maps.c:
```diff
@@ -0,0 +1,528 @@
+
+static DEFINE_MUTEX(mshv_local_maps_mutex);
+
+#define MAX_HYPERV_CPUS 			(u64)(2048)
```
Does num of CPUs ever go beyond 2048? If this limit is defined somewhere else in Hyper-V code, we should use that.

namancse replied:
include/hyperv/hvgdk_mini.h
#if defined(CONFIG_X86)
#define HV_MAXIMUM_PROCESSORS       2048

romank-msft replied:
Good call, fixing that! Thank you for catching this 😄

namancse commented on drivers/hv/mshv_vtl_local_maps.c:
```diff
@@ -0,0 +1,528 @@
+#define NORMAL_PAGE_SIZE 			(u64)(PAGE_SIZE)
+#define ENTRIES_PER_PGTABLE			(u64)(0x200)
+
+#if CONFIG_PGTABLE_LEVELS < 5
```
Adding a comment with memory map diagram can help visualize how you are using it. Not a major concern though.

romank-msft replied:
I'll start on that, leaving the comment open

namancse commented on drivers/hv/mshv_vtl_main.c:
```diff
@@ -1571,36 +1601,62 @@ static long mshv_vtl_ioctl_pvalidate(void __user *pval_user)
+		u64 count = 0;
+		u64 failed_pfn = -1;
+		
+		if (use_large_page) {
```
I see 'count' is being used to keep track of pages pvalidated in a loop. However, I did not understand how below if else code is working. Can you please add a comment if it helps.

romank-msft replied:
Added comments. TLDR; the strategy in the loop is to process as few small pages as possible. The math is to always land on the beginning of a large page if possible

namancse commented on drivers/hv/mshv_vtl_main.c:
```diff
@@ -1571,36 +1601,62 @@ static long mshv_vtl_ioctl_pvalidate(void __user *pval_user)
+		if (use_large_page) {
+			count = ALIGN_DOWN(pfn_end, 0x200) - pfn;
+		} else {
+			if (pfn_end - pfn >= 0x200)
```
In what scenario we would land here? When IS_ALIGNED(pfn, 0x200) is false. But what does it mean, in real terms, meaning when would that happen?

romank-msft replied:
`0x200` is the number of small pages in the large page. ` IS_ALIGNED(pfn, 0x200) is false` means not aligned on the large page boundary. Added comments, replaced the magic number with a macro.

namancse commented on drivers/hv/mshv_vtl_main.c:
```diff
@@ -1571,36 +1601,62 @@ static long mshv_vtl_ioctl_pvalidate(void __user *pval_user)
-			vaddr = kmap_local_page(pfn_to_page(pfn));
-		else
-			vaddr = vmap_pfn(pfns, ARRAY_SIZE(pfns), PAGE_KERNEL);
+		pr_debug("%s: large_page %d, pfn %#llx, count %#llx\n", __func__, use_large_page, pfn, count);
```
IMO, we would be printing this 2048 times, when loglevel allows us to. We can limit printing this info to the case when it fails for some reason. Ditto for other debug prints.

romank-msft replied:
Right, maybe I've went a bit overboard with diagnostics here. Removed

namancse commented on drivers/hv/mshv_vtl_main.c:
```diff
@@ -1571,36 +1601,62 @@ static long mshv_vtl_ioctl_pvalidate(void __user *pval_user)
+
+		rc = mshv_use_local_page(pfn, use_large_page, count, &failed_pfn, mshv_vtl_use_local_page_pvalidate, &param);
+		if (rc == PVALIDATE_FAIL_SIZEMISMATCH && use_large_page) {
+			BUG_ON(!IS_ALIGNED(pfn, 0x200) || (pfn_end - pfn < 0x200));
```
I think, as my previous comment also suggested, I did not understand these checks completely. May be a comment would help?

romank-msft replied:
```c
			/*
			 * The hypervisor indicated that it smashed the large page into 4KiB ones.
			 * That may happen if and only if large pages are used. Assert that is true,
			 * as otherwise something fundamental is broken and the processing has to stop
			 * as the results are undefined.
			 */
```
added

namancse commented on drivers/hv/mshv_vtl_main.c:
```diff
@@ -1571,36 +1601,62 @@ static long mshv_vtl_ioctl_pvalidate(void __user *pval_user)
+			param.use_large_page = false;
+
+			pr_debug("%s: retrying, large_page %d, pfn %#llx, count %#llx\n", __func__, use_large_page, pfn, count);
+			rc = mshv_use_local_page(pfn, use_large_page, count, &failed_pfn, mshv_vtl_use_local_page_pvalidate, &param);
```
is this like retrying with regular pages when large page pvalidate failed?

romank-msft replied:
Exactly! If the hypervisor needs to smash the large page into small ones, it return "size mismatch", and the guest retries.

namancse commented on drivers/hv/mshv_vtl_main.c:
```diff
@@ -1623,34 +1679,57 @@ static long mshv_vtl_ioctl_rmpadjust(void __user *rmpa_user)
 
 	pfn = rmpa.start_pfn;
 	pfn_end = pfn + rmpa.page_count;
-
```
These two pvalidate and rmpadjust functions have duplicate code. If it's easy to squash them into one, that would look better.

romank-msft replied:
Agreed, will attempt refactoring

romank-msft replied:
I hesitated as that would prevent the compiler from optimizing I'm remembering as would have to call the function (pvalidate and rmpadjust wrappers by the pointer) thus not allowing inlining. Perhaps I am optimizing prematurely and the maintainability is a better goal.

namancse replied:
It's fine either way, however you prefer.

namancse commented on drivers/hv/mshv_vtl_main.c:
```diff
@@ -1623,34 +1679,57 @@ static long mshv_vtl_ioctl_rmpadjust(void __user *rmpa_user)
-		else
-			vaddr = vmap_pfn(pfns, ARRAY_SIZE(pfns), PAGE_KERNEL);
+		pr_debug("%s: large_page %d, pfn %#llx, count %#llx\n", __func__, use_large_page, pfn, count);
+		BUG_ON(!count || count > rmpa.page_count);
```
In what scenarios would we hit this BUG_ON?

romank-msft replied:
When the math above in the if statement is broken. We must have more than one and fewer than 0x200 pages.

namancse commented on drivers/hv/mshv_vtl_main.c:
```diff
@@ -1623,34 +1679,57 @@ static long mshv_vtl_ioctl_rmpadjust(void __user *rmpa_user)
+		param.attrs = rmpa.value;
+
+		rc = mshv_use_local_page(pfn, use_large_page, count, &failed_pfn, mshv_vtl_use_local_page_rmpadjust, &param);
+		if (rc == PVALIDATE_FAIL_SIZEMISMATCH && use_large_page) {
```
I hope this is not a typo since rc suggests a error code for pvalidate in rmp operation.

romank-msft replied:
Thanks for raising that. All good, `rmpadjust` and `pvalidate` share error codes. AMD defined the constants this way in the kernel...

namancse commented on drivers/hv/mshv_vtl_local_maps.c:
```diff
@@ -0,0 +1,528 @@
+			vaddr += ENTRIES_PER_PGTABLE*NORMAL_PAGE_SIZE;
+		else
+			vaddr += ENTRIES_PER_PGTABLE*LARGE_PAGE_SIZE;
+		pr_debug("%s: vaddr %#llx\n", __func__, vaddr);
```
Will this lead to excessive logging? Ditto for all other debug logs.

romank-msft replied:
Removed the chattiest ones

namancse commented on drivers/hv/mshv_vtl_local_maps.c:
```diff
@@ -0,0 +1,528 @@
+	mutex_unlock(&mshv_local_maps_mutex);
+}
+
+void mshv_vtl_dump_local_maps(struct mshv_local_maps *maps)
```
Where are we using this function? Am I missing something?

romank-msft replied:
For debugging purposes only 😄

namancse commented on drivers/hv/mshv_vtl_local_maps.c:
```diff
@@ -0,0 +1,528 @@
+
+		if (large) {
+			if (IS_ALIGNED(pfn, ENTRIES_PER_PGTABLE*ENTRIES_PER_PGTABLE))
+				pet_watchdogs();
```
you may want to reconsider placement of this pet_watchdogs(). If this is what you intended to do, please add a comment why we are doing it here. In general, timing is one important factor in deciding "when" to pet a watchdog so that we are not too late. So if we want to do that in every loop iteration, we can relocate it.

romank-msft replied:
That is not done every loop iteration, only every 512th iteration which `IS_ALIGNED` takes care of. Will add a comment

namancse reviewed (changes requested):
Hi Roman, I have reviewed the changes. I am new to MSHV code, so some comments might be very basic. Overall, the problem you are solving is quite interesting and the improvements are really great.

romank-msft commented:
> Hi Roman, I have reviewed the changes. I am new to MSHV code, so some comments might be very basic. Overall, the problem you are solving is quite interesting and the improvements are really great.

Hi Naman, I really appreciate you taking time to dig in into the code! No worries, I take all questions as a signal to improve the code or comments and perhaps reevaluate the assumptions and the model. I might not be immediately able to go through your feedback today; hoping to get back to it this week.

Thanks :)

romank-msft commented:
Hey @namancse, I'm getting to this today or tomorrow! It's been a while, mayve you'd like to give that another look.

romank-msft committed:
Drivers: hv: mshv_vtl: Speed up PVALIDATE and RPMADJUST to meet perf targets

Using 4K pages for pvalidate breaks large VMs in the cloud as that
is very slow. Using 2M pages for mapping arbitrary PFNs (i.e. not
having a tracking struct page*) isn't obvious in the kernel.

Employ the VA gap documented as meant for the hypervisor use for creating local maps.
These are local to the processor and are used to briefly map the PFN requested to
be pvalidate-d or rmpadjust-ed.

The implementation allows to parallelize the load between the CPUs if the user mode
code wishes to do so. The 5-level page tables are supported as well as 4-level ones.

The 64GiB+64VP VMs accept the memory in under a second on average with this change,
and without that that's ~57sec on average.

Signed-off-by: Roman Kisel <romank@linux.microsoft.com>

romank-msft commented:
Validated with the latest atm OpenVMM code.

namancse reviewed (approved):
Changes looks good to me. Thanks Roman for addressing the comments.

romank-msft commented:
> Changes looks good to me. Thanks Roman for addressing the comments.

Thanks again for investing so much in our shared success!!

romank-msft commented:
Validated again with the latest OpenVMM trre, works. Merging
